File: 2024/day-08/main.lua
function table.count(t)
    local count = 0
    for _ in pairs(t) do
        count = count + 1
    end
    return count
end


local function manhattan_distance(r1, c1, r2, c2)
    return math.abs(r1 - r2) + math.abs(c1 - c2)
end

local infile = arg[1] or "input.txt"
local file = io.open(infile, "r")
local grid = {}

for line in file:lines() do
    table.insert(grid, line)
end
file:close()

local R = #grid
local C = #grid[1]
local positions = {}

-- Store antenna positions
for r = 1, R do
    for c = 1, C do
        local char = grid[r]:sub(c, c)
        if char ~= "." then
            positions[char] = positions[char] or {}
            table.insert(positions[char], {r = r, c = c})
        end
    end
end

local a1 = {}
local a2 = {}

for r = 1, R do
    for c = 1, C do
        for _, pos in pairs(positions) do
            for _, p1 in ipairs(pos) do
                for _, p2 in ipairs(pos) do
                    if p1 ~= p2 then
                        local d1 = manhattan_distance(r, c, p1.r, p1.c)
                        local d2 = manhattan_distance(r, c, p2.r, p2.c)
                        local dr1, dc1 = r - p1.r, c - p1.c
                        local dr2, dc2 = r - p2.r, c - p2.c

                        if (d1 == 2 * d2 or d1 * 2 == d2) and dr1 * dc2 == dc1 * dr2 then
                            a1[r * C + c] = true
                        end
                        if dr1 * dc2 == dc1 * dr2 then
                            a2[r * C + c] = true
                        end
                    end
                end
            end
        end
    end
end

print(table.count(a1))
print(table.count(a2))

File: 2024/day-08/main.dart
import 'dart:io';
import 'dart:math';

int manhattanDistance(int r1, int c1, int r2, int c2) {
  return (r1 - r2).abs() + (c1 - c2).abs();
}

void main(List<String> args) {
  String infile = args.isNotEmpty ? args[0] : 'input.txt';
  List<String> grid = File(infile).readAsLinesSync();

  int r = grid.length;
  int c = grid[0].length;

  Map<String, List<List<int>>> positions = {};
  for (int i = 0; i < r; i++) {
    for (int j = 0; j < c; j++) {
      String char = grid[i][j];
      if (char != '.') {
        positions.putIfAbsent(char, () => []).add([i, j]);
      }
    }
  }

  Set<List<int>> a1 = {};
  Set<List<int>> a2 = {};

  for (int i = 0; i < r; i++) {
    for (int j = 0; j < c; j++) {
      positions.forEach((_, pos) {
        for (var p1 in pos) {
          for (var p2 in pos) {
            if (p1 != p2) {
              int d1 = manhattanDistance(i, j, p1[0], p1[1]);
              int d2 = manhattanDistance(i, j, p2[0], p2[1]);
              int dr1 = i - p1[0];
              int dr2 = i - p2[0];
              int dc1 = j - p1[1];
              int dc2 = j - p2[1];

              if ((d1 == 2 * d2 || d1 * 2 == d2) && dr1 * dc2 == dc1 * dr2) {
                a1.add([i, j]);
              }
              if (dr1 * dc2 == dc1 * dr2) {
                a2.add([i, j]);
              }
            }
          }
        }
      });
    }
  }

  print(a1.length);
  print(a2.length);
}
